qiskit_ibm_provider/transpiler/passes/scheduling/block_base_padder.py

```diff
@@ -79,57 +102,69 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
-                )
-
-            self._prev_node = node
-
-        # terminate final block
-        self._terminate_block(self._block_duration, self._current_block_idx, None)
-
-        return self._dag
+        return new_dag
 
     def _init_run(self, dag: DAGCircuit) -> None:
         """Setup for initial run."""
         self._node_start_time = self.property_set["node_start_time"].copy()
+        self._node_block_dags = self.property_set["node_block_dags"]
         self._idle_after = {bit: 0 for bit in dag.qubits}
         self._current_block_idx = 0
         self._conditional_block = False
         self._block_duration = 0
 
         # Prepare DAG to pad
-        self._dag = DAGCircuit()
-        for qreg in dag.qregs.values():
-            self._dag.add_qreg(qreg)
-        for creg in dag.cregs.values():
-            self._dag.add_creg(creg)
-
-        # Update start time dictionary for the new_dag.
-        # This information may be used for further scheduling tasks,
-        # but this is immediately invalidated because most node ids are updated in the new_dag.
+        self._root_dag = dag
+        self._dag = self._empty_dag_like(dag)
+        self._block_dag = self._dag
+        self._bit_indices = {q: index for index, q in enumerate(dag.qubits)}
+        self._last_node_to_touch = {}
+        self._fast_path_nodes = set()
+        self._dirty_qubits = set()
+
         self.property_set["node_start_time"].clear()
+        self._prev_node = None
 
-        self._dag.name = dag.name
-        self._dag.metadata = dag.metadata
-        self._dag.unit = self.property_set["time_unit"]
-        self._dag.calibrations = dag.calibrations
-        self._dag.global_phase = dag.global_phase
+    def _empty_dag_like(self, dag: DAGCircuit, pad_wires: bool = True) -> DAGCircuit:
+        """Create an empty dag like the input dag."""
+        new_dag = DAGCircuit()
 
-        self._prev_node = None
+        # Ensure *all* registers are included from the input circuit
+        # so that they are scheduled in sub-blocks
+
+        # Control flow blocks do not get the full reg added to the
+        # block but just the bits. To work around this we try to
+        # add the reg if available and otherwise add the bits directly.
+        # We need this work around as otherwise the padded circuit will
+        # not be equivalent to one written manually as bits will not
+        # be defined on registers like in the test case.
+
+        source_wire_dag = self._root_dag if pad_wires else dag
+
+        if dag.qregs:
+            for qreg in source_wire_dag.qregs.values():
+                new_dag.add_qreg(qreg)
+        else:
+
+            new_dag.add_qubits(source_wire_dag.qubits)
+
+        # Don't add root cargs as these will not be padded.
+        # Just focus on current block dag.
+        if dag.cregs:
+            for creg in dag.cregs.values():
+                new_dag.add_creg(creg)
+        else:
+            new_dag.add_clbits(dag.clbits)
```

Comment:
Unfortunately, this doesn't help because this would add the registers to the nested blocks. The problem is actually with the hand-written circuit builder equivalence circuits which *don't* have registers. This seems like an either-or scenario.

Reply:
I'm not sure I see what you mean - if `dag` doesn't have cregs then the for loop is a no-op, while if it does (and also if it has some loose clbits _and_ some registers) then my form will add them too, without needing the if/else.

Reply:
I've updated to your code for the clbits in 812de1f. However, I maintain this does not work for qubits as nested blocks do not receive the qregs - the source wire dag _will_ have the qregs as it is a "top-level" block.

This code enables us to handle both the top-level and "nested" QuantumCircuit "blocks". Which IMO should be disambiguate by a "Program/Block" distinction. (where a program has a single entry Quantum Circuit block).